src/api/eth_subscribe.rs

```diff
@@ -0,0 +1,161 @@
+//! `Eth` namespace, subscriptions
+
+use std::fmt;
+use std::marker::PhantomData;
+
+use api::Namespace;
+use futures::{Async, Future, Poll, Stream};
+use helpers::{self, CallResult};
+use rpc;
+use serde;
+use serde_json;
+use types::{BlockHeader, Filter, H256, Log};
+use {DuplexTransport, Error};
+
+/// `Eth` namespace, subscriptions
+#[derive(Debug, Clone)]
+pub struct EthSubscribe<T> {
+  transport: T,
+}
+
+impl<T: DuplexTransport> Namespace<T> for EthSubscribe<T> {
+  fn new(transport: T) -> Self
+  where
+    Self: Sized,
+  {
+    EthSubscribe { transport }
+  }
+
+  fn transport(&self) -> &T {
+    &self.transport
+  }
+}
+
+pub struct SubscriptionStream<T: DuplexTransport, I> {
+  transport: T,
+  id: String,
```

Comment:
Can we extract `Id` to a typedef? Will be easier to read than a bunch of `String`s

Reply:
Actually maybe best to have a wrapper type:
```
#[derive(...)]
pub struct Id(String)
```
to make all interactions with id typesafe.